### DISCUSSION TITLE
2.2.0

### DISCUSSION THREAD

**Discussion by taitus** [consuldemocracy/consuldemocracy#5591] [Releases]
Here's release **2.2.0** with:

* Configurable default and enabled languages using the admin section
* A map displaying GeoJSON coordinates for geozones
* Configurable order for the homepage cards
* Better GDPR compliance by removing default tracking-related cookies
* Maintenance updates like upgrading to Rails 7.0 and Ruby 3.2
* Many bug fixes, accessibility enhancements and small improvements

Check the [Changelog](https://github.com/consuldemocracy/consuldemocracy/blob/master/CHANGELOG.md#210-2024-07-01) to see all the changes. Please take your time reading all referenced pull requests to understand new features and changes.

## Updating your fork

Check [how to keep your fork updated](https://docs.consuldemocracy.org/tech_docs/getting_started/update#pulling-changes-from-consul-democracy); if you're updating from release 2.1.1, you can just:

```
git checkout master
git pull
git checkout -b release
git fetch upstream tag 2.2.0
git merge 2.2.0
git checkout master
git merge release
git branch -d release
git push
```

:gear: After deploying the new version, execute the release tasks:

```
bin/rake consul:execute_release_tasks RAILS_ENV=production
```

## Release Notes
:warning: We've changed the Rails autoloader to Zeitwerk, which is the only autoloader that works with Rails 7. **If you use `require_dependency` in your custom code** (**and, if you use custom `.rb` files, you most likely do**, since it's the system we've mentioned in our documentation for years) **you'll have to replace it with `load`**. See commit 6552e3197 from pull request #5425 for an example showing how to do so.

:warning: We've moved most of the content of the `lib/` folder to the `app/lib/` folder. If you've got custom code in the `lib/` folder, you'll have to move it to `app/lib/` as well unless you're using `require` to load it where it's needed.

:warning: We're [replaced Libsass with Dart Sass](https://github.com/consuldemocracy/consuldemocracy/pull/5477). Since we were getting a couple of syntax warnings (caused by external dependencies) every time we changed a SCSS file, we've disabled these warnings. If you've written custom SCSS code, you might want to temporarily enable them in development to check your custom code; you can do that by removing the `config/initializers/sass.rb` file.

:information_source: Due to a race condition, until now, there was about a 1 in 10,000 chance to get duplicate records when people voted in a poll. This release fixes that (see pull requests #5532 and #5539) and, as part of the release tasks, provides tasks to remove the duplicate records. A log file `log/duplicate_records.log` will contain the information about the deleted records. As usual, backup your database before upgrading.

:warning: In order to try to help developers to avoid losing their sanity, we've renamed the `Poll::Question::Answer` model to `Poll::Question::Option`. If you've got some custom codes dealing with polls, you might have to update it. For now, we are **not** renaming the related database tables in order to make the upgrade easier.

:warning: We've added a new `option_id` field to the `poll_answers` table; a task that populates this field is part of the release tasks. If you've added custom votation types to your code, you might need to modify this task. Also, this task doesn't handle an edge case where the options for a question were renamed after the poll started (which isn't possible since pull request #4904) and an extremely rare case that only applies to multi-language polls and that we're not even sure is possible in practice. For either of these cases, you'll get a warning when running the task, so you can choose what to do about it (you should handle these cases, since future version will use the `option_id` column to calculate results and stats). The task doesn't change records which already have an `option_id`, so it can be run as many times as you'd like to. Also note that, if you've got custom code to save records in the `poll_answers` table, you'll have to make sure that, from now on, that code populates both the `answer` column and the `option_id` column (unless you've implemented open answers in your custom code; in that case, you can skip the `option_id` when dealing with open answers).

:warning: We've removed the code used to check the `campaigns` table. It's extremely unlikely that you have any data in the `campaigns` table because it could only be introduced manually in the database but if, for some reason, you have information there, you'll have to copy it to a different table because we'll remove the `campaigns` table in version 2.3.0.

:warning: We've also removed the code related to Ahoy events, since we only used it to generate charts and their data wasn't accurate (see pull request #5503). Since we're no longer using the information stored there, we'll remove the `ahoy_events` table in version 2.3.0 so, if you're using this table for other purposes, make sure to copy its data to a different table.

<hr /><em>This discussion was created from the release <a href='https://github.com/consuldemocracy/consuldemocracy/releases/tag/2.2.0'>2.2.0</a>.</em>

**Comment by jpprost** [comment]
Hello,

as I'm trying to upgrade to release 2.2.0 from 2.1.1 the deployment fails, seemingly on a localisation problem.
The tail of the output on stdout is here:
https://nimbus.jpprost.fr/s/32ptJK2My5sBeLN

The detailed capistrano log file is here:
https://nimbus.jpprost.fr/s/7A2TeJxXkWS2GHT

any suggestions?

The following message
```
Caused by:
Psych::SyntaxError: (/home/deploy/consul/releases/20240728155200/config/locales/custom/fr/general.yml): did not find expected key while parsing a block mapping at line 2 column 3 (Psych::SyntaxError)
```
makes it look to me like it's not something I can figure out by myself.

Cheers,
JP

=== COMPLEMENT, 29 July 24:
the GitHub CI test suite failed after the merge:
![image](https://github.com/user-attachments/assets/6f530127-ccd9-400a-82cc-ec4058497aac)

**Comment by javierm** [comment]
Hi, @jpprost :smile:.

Looks like your `config/locales/custom/fr/general.yml` contains a syntax error or an interpolation issue :thinking:. Could you paste a link to your repository?

**Reply by jpprost** [reply]
Hi @javierm!

Hmmm... I didn't realise it was coming from a custom file, sorry.
Here's my repo:
https://github.com/LaRoqueEnCommun/consuldemocracy

**Reply by javierm** [reply]
It looks like the `confirm_delete` key is using 4 spaces for indentation instead of 2, and that's causing the syntax error.

By the way, in case it helps, I've noticed this key is under `application`, while it should probably be under `comments.actions` :thinking:. Same happens with some other keys; for example, `comment` should probably be under `comments` instead of being under `application`.

Hope that helps!

**Reply by jpprost** [reply]
Hmmm... ok, thank you. I didn't realise that yaml localisation files could make tests fail.
Cheers,
JP

**Reply by javierm** [reply]
> I didn't realise that yaml localisation files could make tests fail.

Thankfully, they do :smile:. This way invalid YAML code will never reach production.

**Reply by jpprost** [reply]
Hi @javierm :-)

I now have a custom `general.yml` that is syntactically correct, and the test suite is sucessful when I push.
However, two problems still occur:

1. After deployment on the production env, the custom texts seem not to be taken into account when I reload the site. I can't explain why not.
It may be worth mentioning that with the previous version some of the custom text was taken into account, despite the syntax errors discussed above.
2. After deployment, the execution of the release tasks as per instructed resulted in the following:
 a) a first failure, with the following error message:
 ```
bin/rake consul:execute_release_tasks RAILS_ENV=production
Could not find rails-7.0.8.4, actioncable-7.0.8.4, actionmailbox-7.0.8.4, actionmailer-7.0.8.4, actionpack-7.0.8.4, actiontext-7.0.8.4, actionview-7.0.8.4, activejob-7.0.8.4, activemodel-7.0.8.4, activerecord-7.0.8.4, activestorage-7.0.8.4, activesupport-7.0.8.4, railties-7.0.8.4, nio4r-2.7.3, concurrent-ruby-1.3.1, i18n-1.14.5, method_source-1.1.0, parser-3.3.1.0, rake-13.2.1, net-imap-0.4.12, minitest-5.23.1, zeitwerk-2.6.15, racc-1.8.0 in locally installed gems
Run `bundle install --gemfile /home/prost/LaREC/consuldemocracy/consuldemocracy/Gemfile` to install missing gems.
```
 b) after running the requested `bundle` command, the process failed again with the following message:
 ```
 rake aborted!
ArgumentError: Missing `secret_key_base` for 'production' environment, set this string with `bin/rails credentials:edit` (ArgumentError)

        raise ArgumentError, "Missing `secret_key_base` for '#{Rails.env}' environment, set this string with `bin/rails credentials:edit`"
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/share/rvm/gems/ruby-3.2.3/gems/railties-7.0.8.4/lib/rails/application.rb:576:in `validate_secret_key_base'
/usr/share/rvm/gems/ruby-3.2.3/gems/railties-7.0.8.4/lib/rails/application.rb:419:in `secret_key_base'
/usr/share/rvm/gems/ruby-3.2.3/gems/devise-4.9.3/lib/devise/secret_key_finder.rb:24:in `key_exists?'
/usr/share/rvm/gems/ruby-3.2.3/gems/devise-4.9.3/lib/devise/secret_key_finder.rb:16:in `find'
/usr/share/rvm/gems/ruby-3.2.3/gems/devise-4.9.3/lib/devise/rails.rb:41:in `block in <class:Engine>'
/usr/share/rvm/gems/ruby-3.2.3/gems/railties-7.0.8.4/lib/rails/initializable.rb:32:in `instance_exec'
/usr/share/rvm/gems/ruby-3.2.3/gems/railties-7.0.8.4/lib/rails/initializable.rb:32:in `run'
/usr/share/rvm/gems/ruby-3.2.3/gems/railties-7.0.8.4/lib/rails/initializable.rb:61:in `block in run_initializers'
/usr/share/rvm/gems/ruby-3.2.3/gems/railties-7.0.8.4/lib/rails/initializable.rb:60:in `run_initializers'
/usr/share/rvm/gems/ruby-3.2.3/gems/railties-7.0.8.4/lib/rails/application.rb:372:in `initialize!'
/home/prost/LaREC/consuldemocracy/consuldemocracy/config/environment.rb:5:in `<top (required)>'
<internal:/usr/share/rvm/rubies/ruby-3.2.3/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
<internal:/usr/share/rvm/rubies/ruby-3.2.3/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:38:in `require'
/usr/share/rvm/gems/ruby-3.2.3/gems/zeitwerk-2.6.15/lib/zeitwerk/kernel.rb:34:in `require'
/usr/share/rvm/gems/ruby-3.2.3/gems/railties-7.0.8.4/lib/rails/application.rb:348:in `require_environment!'
/usr/share/rvm/gems/ruby-3.2.3/gems/railties-7.0.8.4/lib/rails/application.rb:506:in `block in run_tasks_blocks'
/usr/share/rvm/gems/ruby-3.2.3/gems/airbrake-13.0.4/lib/airbrake/rake.rb:17:in `execute'
Tasks: TOP => consul:execute_release_tasks => settings:rename_setting_keys => environment
(See full trace by running task with --trace)
```

Any suggestion? Could the two problems be related with each other?

Cheers,
JP

**Reply by jpprost** [reply]
**UPDATE**

I was not working with the correct version of Ruby -- the problem being that rvm only checks that the required version is the one in use when changing directory on the terminal, which I had forgotten... My mistake, sorry (though if I may, that is quite unusual a requirement!).
Now I am running the correct version, and the test suite fails again, on yet a different issue. Here is the message (only the core):

```
Failures:

  1) Welcome screen behaves like remotely_translatable After click remote translations button without delayed jobs request a translation of an already translated text
     Failure/Error: expect(page).to have_content "Se han solicitado correctamente las traducciones"
       expected to find text "Se han solicitado correctamente las traducciones" in "Idioma: \n        \nEnglish\nDeutsch\nEspañol\nFrançais\nNederlands\nPortuguês brasileiro\n中文\n         Groupe d'Intérêt Communal LaREC\nEntrar\nRegistrarse\nDebates\nPropuestas\nVotaciones\nLegislación colaborativa\nPresupuestos participativos\nODS\nAyuda\n×\nTranslations have been correctly requested.\nPropuestas más activas\nAhora mismo no hay propuestas\nDebates más activos\nzq9zaafl53\nVer todos los debates\nProcesos abiertos\nAhora mismo no hay procesos abiertos\nGobierno abierto\nEste portal usa la aplicación CONSUL DEMOCRACY que es software de código abierto.\nParticipación\nDecide cómo debe ser la ciudad que quieres.\nCONSUL DEMOCRACY, 2024 Política de privacidad Condiciones de uso Accesibilidad"

     [Screenshot Image]: /home/runner/work/consuldemocracy/consuldemocracy/tmp/capybara/failures_r_spec_example_groups_welcome_screen_behaves_like_remotely_translatable_2_after_click_remote_translations_button_without_delayed_jobs_request_a_translation_of_an_already_translated_text_247.png


     Shared Example Group: "remotely_translatable" called from ./spec/system/welcome_spec.rb:7
     # ./spec/shared/system/remotely_translatable.rb:218:in `block (5 levels) in <top (required)>'
     # ./spec/sessions_helper.rb:2:in `in_browser'
     # ./spec/shared/system/remotely_translatable.rb:213:in `block (4 levels) in <top (required)>'
     # ./spec/spec_helper.rb:40:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:39:in `block (2 levels) in <top (required)>'

  2) Moderate proposal notifications Hide
     Failure/Error:
       within("#proposal_notification_#{proposal_notification.id}") do
         accept_confirm("Are you sure? Hide") { click_button "Hide" }
       end

     Capybara::ElementNotFound:
       Unable to find visible css "#proposal_notification_20"

     [Screenshot Image]: /home/runner/work/consuldemocracy/consuldemocracy/tmp/capybara/failures_r_spec_example_groups_moderate_proposal_notifications_hide_893.png


     # ./spec/system/moderation/proposal_notifications_spec.rb:16:in `block (2 levels) in <top (required)>'
     # ./spec/spec_helper.rb:40:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:39:in `block (2 levels) in <top (required)>'

  3) Moderate proposal notifications Can not hide own proposal notification
     Failure/Error:
       within("#proposal_notification_#{proposal_notification.id}") do
         expect(page).not_to have_button "Hide"
         expect(page).not_to have_button "Block author"
       end

     Capybara::ElementNotFound:
       Unable to find visible css "#proposal_notification_21"

     [Screenshot Image]: /home/runner/work/consuldemocracy/consuldemocracy/tmp/capybara/failures_r_spec_example_groups_moderate_proposal_notifications_can_not_hide_own_proposal_notification_551.png


     # ./spec/system/moderation/proposal_notifications_spec.rb:40:in `block (2 levels) in <top (required)>'
     # ./spec/spec_helper.rb:40:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:39:in `block (2 levels) in <top (required)>'

Finished in 7 minutes 22 seconds (files took 7.25 seconds to load)
1123 examples, 3 failures

Failed examples:

rspec './spec/system/welcome_spec.rb[1:2:2:2:3]' # Welcome screen behaves like remotely_translatable After click remote translations button without delayed jobs request a translation of an already translated text
rspec ./spec/system/moderation/proposal_notifications_spec.rb:4 # Moderate proposal notifications Hide
rspec ./spec/system/moderation/proposal_notifications_spec.rb:29 # Moderate proposal notifications Can not hide own proposal notification

```

Re failure 1), I'm quite surprised because the feature "Remote translation" was turned off. However, the "Translation interface" was on (I was curious to see what it offers).

Turning off the "Translation interface" fixes the failure 1) above. However, the custom localisation file still isn't taken into account.
For the time being, I will provide my own translations through the admin>site content>custom information texts menu.

Although the initial problem seems partially solved, and an acceptable workaround exists for the custom translation issue, I leave my comments here for information's sake, in case an explanation can be found.

Thank you!

Cheers,
JP

**Reply by javierm** [reply]
Hi, @jpprost :smile:.

> the custom texts seem not to be taken into account when I reload the site

Thanks for reporting! It looks like this feature stopped working when we upgraded to Rails 7.0. I've just opened pull request #5660 to fix it.

> It may be worth mentioning that with the previous version some of the custom text was taken into account, despite the syntax errors discussed above.

Since the application didn't even load with the syntax errors and the deployment failed, it's likely that the production server was using a previous version which didn't have the syntax errors.

> `bin/rake consul:execute_release_tasks RAILS_ENV=production`

On which machine are you running this command? This should be run on the production machine, under the current release folder.

> Turning off the "Translation interface" fixes the failure 1) above. However, the custom localisation file still isn't taken into account.

The test database is not affected by changes on the development or production databases, so I think this is a coincidence :thinking:. Might be a case of "flaky tests"; unfortunately, some tests fail under certain conditions that aren't always easy to reproduce, either due to a rare bug in the application or due to a rare bug in the tests :relieved: (also unfortunately, this is an extremely common problem in software development, particularly with system tests), although the tests you mention have never failed for us AFAIK :thinking:.

In any case, I see your CI is already passing :ok_hand:.

> I'm quite surprised because the feature "Remote translation" was turned off.

This is a feature that enables integration with microsoft translator, and it's turned off by default because you need a microsoft translator api key in order to be able to use it. Maybe we should add it to the feature description (pull requests are welcome!).

**Reply by jpprost** [reply]
Hi @javierm!

thank you for your reply.

> > `bin/rake consul:execute_release_tasks RAILS_ENV=production`
> 
> On which machine are you running this command? This should be run on the production machine, under the current release folder.

Ah! Ok. Silly me. It runs fine on the production server.

> (pull requests are welcome!).

I know... I must get familiar with the process. I am planning to.

Cheers,
JP
